sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala

```diff
@@ -52,16 +53,25 @@ case class CollectLimitExec(limit: Int, child: SparkPlan) extends LimitExec {
     SQLShuffleReadMetricsReporter.createShuffleReadMetrics(sparkContext)
   override lazy val metrics = readMetrics ++ writeMetrics
   protected override def doExecute(): RDD[InternalRow] = {
-    val locallyLimited = child.execute().mapPartitionsInternal(_.take(limit))
-    val shuffled = new ShuffledRowRDD(
-      ShuffleExchangeExec.prepareShuffleDependency(
-        locallyLimited,
-        child.output,
-        SinglePartition,
-        serializer,
-        writeMetrics),
-      readMetrics)
-    shuffled.mapPartitionsInternal(_.take(limit))
+    val childRDD = child.execute()
+    if (childRDD.getNumPartitions == 0) {
+      new ParallelCollectionRDD(sparkContext, Seq.empty[InternalRow], 1, Map.empty)
```

Comment:
EmptyRDD?

Reply:
kindly ping @zhenglaizhang

Reply:
@viirya @maropu  we can not set the number of partitions in an `EmptyRDD`, it always has zero partitions.

Reply:
Isn't the `childRDD` zero partitions?

Reply:
`new ParallelCollectionRDD(sparkContext, Seq.empty[InternalRow], 1, Map.empty)` is a empty rdd with single partition

Reply:
Oh I see. I am not sure if `CollectLimitExec` must be single partition, but this looks minor. I'm okay with `ParallelCollectionRDD`. Thanks.

Reply:
@viirya the `outputPartitioning` of `CollectLimitExec` is  `SinglePartition`, so the output rdd should have single partition.

Reply:
or we can use `EmptyRDDWithPartitions` defined in `CoalesceExec`?
maybe we can make `EmptyRDD` support number of partition, so we can use it in both `CollectLimitExec` and `CoalesceExec`. But I am not sure whether we should do this.

Reply:
> @viirya the `outputPartitioning` of `CollectLimitExec` is `SinglePartition`, so the output rdd should have single partition.

Oh, my previous comment was confusing. I mean I am not sure if its `outputPartitioning` must be ` SinglePartition`.

Reply:
> or we can use `EmptyRDDWithPartitions` defined in `CoalesceExec`?
> maybe we can make `EmptyRDD` support number of partition, so we can use it in both `CollectLimitExec` and `CoalesceExec`. But I am not sure whether we should do this.

I think it sounds over-engineering. At least for now I don't think we need it.